---
title: "Cycle model: External data"
author: "Jessica Chung"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: true
    toc_depth: 4
---

Endo data from GSE141549  
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE141549

A relational database to identify differentially expressed genes in the endometrium and endometriosis lesions  
https://www.nature.com/articles/s41597-020-00623-x

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=TRUE, message=TRUE, error=TRUE, echo=TRUE, results="hold")
knitr::opts_knit$set(root.dir = "..")
options(digits=4)
options(width=120)
Sys.setenv(TZ = "Australia/Melbourne")
```

-----

# Load

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(limma)
library(ggplot2)
library(sva)
library(splines)
library(readxl)
library(illuminaHumanv3.db)
library(DBI)
library(janitor)
library(preprocessCore)
select <- dplyr::select
```


```{r}
unique_id <- "2021-03-23"
combat_phenotype <- readRDS(here::here(sprintf("data/tidy_data/combat_phenotype_%s.rds", unique_id)))
combat_nc <- readRDS(here::here("data/tidy_data/batch_normalised_exprs.rds"))
# 
unique_id <- "2021-03-23_k30_g4"
gene_models <- readRDS(here::here(sprintf("data/gene_models/gene_models_%s.rds", unique_id)))
```

```{r}
# Get expected_exprs
cycle_range <- gene_models[[1]]$pred$cycle_time
genes <- names(gene_models)
expected_exprs <- sapply(genes, function(p) gene_models[[p]]$pred$pred) %>% 
  t %>% magrittr::set_colnames(paste0("time_", cycle_range))
```


```{r}
all_probe_annotation <- illuminaHumanv3fullReannotation() %>%
  dplyr::select(IlluminaID, ProbeQuality, SymbolReannotated, EnsemblReannotated,
                EntrezReannotated) %>% janitor::clean_names()
```

```{r}
gse_xlsx_filename <- here::here("data/ext_data/GSE141549_batchCorrectednormalizedArrayscombined.xlsx")
gse_data <- read_excel(gse_xlsx_filename) %>% janitor::clean_names()
```

```{r}
gse_1 <- read_delim(here::here("data/ext_data/GSE141549-GPL10558_series_matrix.txt"), delim="\t", comment="!Series") %>%
  janitor::clean_names()
gse_2 <- read_delim(here::here("data/ext_data/GSE141549-GPL13376_series_matrix.txt"), delim="\t", comment="!Series") %>%
  janitor::clean_names()
```


-----

# Wrangle GSE data

### Sample info data

```{r}
stopifnot(gse_1$sample_title == gse_2$sample_title)
gse_phenotype_combined <- cbind(gse_1,  gse_2[,-1])
```

```{r}
gse_phenotype <- gse_phenotype_combined %>%
  filter(sample_title %in% c("!Sample_geo_accession", "!Sample_source_name_ch1",
                          "!Sample_description", "!Sample_characteristics_ch1"))
rownames(gse_phenotype) <- c("geo_accession", "source", "tissue", "age",
                             "disease_stage", "cycle_phase", "description")

gse_phenotype <- gse_phenotype %>% t %>% data.frame(stringsAsFactors=FALSE)
gse_phenotype <- gse_phenotype[-1,] %>% tibble::rownames_to_column("sample_id")

gse_phenotype <- gse_phenotype %>%
  mutate(age=as.numeric(str_remove(age, "age: ")),
         disease_stage=str_remove(disease_stage, "disease stage: "),
         cycle_phase=str_remove(cycle_phase, "cycle phase: "),
         tissue=str_remove(tissue, "tissue: "))
```

### Expression data

```{r}
# all_gse_matrix <- gse_data %>% select(-id_ref, -transcript, -ilmn_gene, -probe_sequence) %>% 
#   as.matrix
all_gse_matrix <- gse_data %>% select(-gene_symbol, -probe_id) %>% 
  as.matrix
rownames(all_gse_matrix) <- gse_data$probe_id
```

```{r}
# gse_data[1:10,1:10]
# all_gse_matrix[1:10,1:10]
probe_info <- all_probe_annotation %>%
  filter(illumina_id %in% gse_data$probe_id,
         ! is.na(ensembl_reannotated),
         str_detect(probe_quality, "Good|Perfect"))

tmp <- probe_info %>% group_by(ensembl_reannotated) %>% summarise(n=n())
table(tmp$n)

single_ensembl <- tmp %>% filter(n == 1) %>% pull(ensembl_reannotated)
multiple_ensembl <- tmp %>% filter(n > 1) %>% pull(ensembl_reannotated)

m <- match(single_ensembl, probe_info$ensembl_reannotated)
single_probes <- probe_info[m, "illumina_id"]
single_exprs <- all_gse_matrix[single_probes,]
rownames(single_exprs) <- single_ensembl

# Consolidate ensembl IDs that have multiple probes
consolidated_exprs <- list()
for (ensembl_id in multiple_ensembl) {
  illumina_ids <- probe_info %>% 
    filter(ensembl_reannotated == ensembl_id) %>% pull(illumina_id)
  consolidated_exprs[[ensembl_id]] <- all_gse_matrix[illumina_ids,] %>% colMeans
}

# Combine single and multiple
gse_matrix <- rbind(
  do.call(rbind, consolidated_exprs),
  single_exprs)

gse_matrix[1:10,1:5]
```


```{r}
common_probes <- intersect(rownames(expected_exprs), rownames(gse_matrix))
length(common_probes)
```


```{r}
filtered_common_probes <- common_probes
```

-----

# Normalisation

```{r}
# Create a reference to quantile normalise against
tmp <- apply(combat_nc, 2, sort)
tmp <- apply(tmp, 1, median)
tmp %>% head

ref <- tmp
head(ref)

gse_norm <- normalize.quantiles.use.target(gse_matrix[filtered_common_probes,], target=ref)
colnames(gse_norm) <- colnames(gse_matrix)
rownames(gse_norm) <- filtered_common_probes
```

-----

# Endo cycle norm

```{r}
get_molecular_time <- function(observed, expected) {
  mse <- sweep(x=expected, MARGIN=1, STATS=observed, FUN="-") %>%
    `^`(2) %>% colMeans
  df <- data.frame(time=as.numeric(str_remove(names(mse), "time_")),
                    mse=mse)
  rownames(df) <- NULL
  time <- df[which.min(df$mse), "time"]
  return(list(mse=mse, mse_df=df, time=time))
}

molecular_model <- function(expression_matrix) {
  sample_names <- colnames(expression_matrix)
  common_genes <- intersect(rownames(expression_matrix), rownames(expected_exprs))
  exp <- expected_exprs[common_genes,]
  results <- lapply(sample_names, function(x) {
    get_molecular_time(observed=expression_matrix[,x], expected=exp)
  })
  mse <- sapply(results, function(x) x$mse)
  colnames(mse) <- sample_names
  rownames(mse) <- str_remove(rownames(mse), "time_")
  model_time <- sapply(results, function(x) x$time)
  names(model_time) <- sample_names
  return(list(mse=mse, time=model_time))
}

plot_mse <- function(object, sample_name="") {
  df <- data.frame(time=as.numeric(rownames(object$mse)),
                   mse=object$mse[,sample_name])
  ggplot(df, aes(x=time, y=mse)) +
    geom_line() +
    geom_vline(xintercept=object$time[sample_name], linetype="dashed", color="red") +
    scale_x_continuous(breaks=seq(0, 100, by=10)) +
    theme_bw() +
    labs(title=sprintf("%s", sample_name), x="Time", y="Mean Square Error")
}
```

```{r}
results <- molecular_model(gse_norm)
```

```{r}
gse_phenotype <- gse_phenotype %>% mutate(model_time=results$time[sample_id])
```

```{r}
endometrium_samples <- gse_phenotype %>% 
  filter(source == "Endometrium") %>%
  select(sample_id, cycle_phase, model_time, description, source, tissue,
         age, disease_stage)
```


```{r}
ggplot(endometrium_samples, aes(x=model_time, y=cycle_phase)) +
  geom_point()
```

```{r}
endometrium_samples %>% filter(cycle_phase %in% c("menstruation", "proliferative", "secretory")) %>%
  ggplot(aes(x=model_time)) +
  geom_histogram(binwidth = 2)
```

```{r}
# MSE plot examples
tmp <- endometrium_samples %>% filter(cycle_phase != "medication") %>% pull(sample_id) %>%
  head(5)
for (samp in tmp) {
  plot_mse(results, samp) %>% print
}
```


```{r}
# Medication MSE plot examples
tmp <- endometrium_samples %>% filter(cycle_phase == "medication") %>% pull(sample_id) %>%
  head(5)
for (samp in tmp) {
  plot_mse(results, samp) %>% print
}
```

```{r}
# Either model or cycle phase is incorrect
tmp <- endometrium_samples %>% filter(model_time < 30 & cycle_phase == "secretory") %>%
  pull(sample_id)
for (samp in tmp) {
  plot_mse(results, samp) %>% print
}
```

-----

# PCA

```{r}
dat <- endometrium_samples %>% filter(cycle_phase %in% c("menstruation", "proliferative", "secretory"))
pca <- prcomp(t(gse_norm[,dat$sample_id]), scale.=FALSE)$x[,1:4]
pca <- as.data.frame(pca) %>% tibble::rownames_to_column("sample_id") %>%
  merge(dat, by="sample_id")

ggplot(pca, aes(x=PC1, y=PC2, color=cycle_phase)) +
  geom_point()
ggplot(pca, aes(x=PC1, y=PC2, color=model_time)) +
  geom_point() +
  scale_colour_gradientn(limits=c(0,100), colors=rainbow(20)[2:19])
```

```{r}
saveRDS(pca, here::here("cache/fig4b.rds"))
```

-----

# Session Info

```{r}
devtools::session_info()
```

